java/pattern/schema/Rule.java

```diff
@@ -0,0 +1,112 @@
+import graql.lang.pattern.Conjunction;
+import graql.lang.pattern.Definable;
+import graql.lang.pattern.Pattern;
+import graql.lang.pattern.variable.ThingVariable;
+
+import java.util.List;
+import java.util.Objects;
+
+import static graql.lang.common.GraqlToken.Char.COLON;
+import static graql.lang.common.GraqlToken.Char.CURLY_CLOSE;
+import static graql.lang.common.GraqlToken.Char.CURLY_OPEN;
+import static graql.lang.common.GraqlToken.Char.SEMICOLON;
+import static graql.lang.common.GraqlToken.Char.SPACE;
+import static graql.lang.common.GraqlToken.Schema.RULE;
+import static graql.lang.common.GraqlToken.Schema.THEN;
+import static graql.lang.common.GraqlToken.Schema.WHEN;
+import static graql.lang.common.exception.ErrorMessage.MISSING_PATTERNS;
+import static java.util.stream.Collectors.toList;
+
+public class Rule implements Definable {
+    private final String label;
+    private Conjunction<? extends Pattern> when;
+    private ThingVariable<?> then;
+    private int hash = 0;
+
+    public Rule(final String label) {
+        this.label = label;
+    }
+
+    @Override
+    public boolean isRule() {
+        return true;
+    }
+
+    @Override
+    public Rule asRule() { return this; }
+
+    public Rule when(final Conjunction<? extends Pattern> when) {
+        if (when == null) throw new NullPointerException("Null when pattern");
+        if (when.patterns().size() == 0) throw GraqlException.of(MISSING_PATTERNS.message());
+        this.when = when;
+        return this;
+    }
+
+    public Rule then(final ThingVariable<?> pattern) {
+        if (pattern == null) throw new NullPointerException("Null then pattern");
+        this.then = pattern;
+        return this;
+    }
+
+    @Override
+    public String toString() {
+        final StringBuilder syntax = new StringBuilder();
+        syntax.append(RULE).append(SPACE).append(label).append(COLON).append(SPACE);
+
+        // when
+        syntax.append(WHEN).append(SPACE).append(CURLY_OPEN).append(SPACE);
+        if (when != null) {
+            for (Pattern p : when.patterns()) {
+                syntax.append(p).append(SEMICOLON).append(SPACE);
+            }
+        }
+        syntax.append(CURLY_CLOSE).append(SPACE);
+
+        // then
+        syntax.append(THEN).append(SPACE).append(CURLY_OPEN).append(SPACE);
+        if (then != null) syntax.append(then).append(SEMICOLON).append(SPACE);
+        syntax.append(CURLY_CLOSE);
+
+        return syntax.toString();
+    }
+
+    @Override
+    public int hashCode() {
+        if (hash == 0) {
+            this.hash = Objects.hash(label, when, then);
+        }
+        return hash;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (o == this) {
+            return true;
+        }
+        if (o instanceof Rule) {
+            final Rule that = (Rule) o;
+            return this.label.equals(that.label) && this.when.equals(that.when) && this.then.equals(that.then);
+        }
+        return false;
```

Comment:
Let's just use the common `equals()` method structure. You missed to check if the `Object o` is `null` here.

Reply:
```
if (this == o) return true;
if (o == null || getClass() != o.getClass()) return false;
final Rule that = (Rule) o;
return (this.label.equals(that.label) && 
            this.when.equals(that.when) && 
            this.then.equals(that.then));
```

Reply:
That's what it should be @flyingsilverfin

Reply:
Why is this marked as resolved, @flyingsilverfin ? Is the change perhaps not pushed yet? Better to resolve comments after they've been pushed.

Reply:
I'm sorry it's GitHub not refreshing the page properly even after I hit refresh 5 times! This has been resolved I can see now.